### ISSUE TITLE
How do I optimise for F1 score?

### ISSUE THREAD

**Issue by umarbutler** [stanfordnlp/dspy#556]
The documentation states that '[f]or simple tasks, [a metric] could be just "accuracy" or "exact match" or "F1 score". This may be the case for simple classification or short-form QA tasks', yet it does not clarify how F1 scores can be used to optimize programs. Is that possible? And if so, how?

The `metric` function that is passed to `teleprompter.compile` seems to take `gold` and `pred` as inputs, which are single classifications, so I am unable to see how you could calculate an F1 score based off of that. It would be really helpful to have that ability since not all tasks should be optimised based on accuracy.

**Comment by umarbutler** [comment]
It would be great if we could define a function that indicates whether a prediction is a true positive, false positive, true negative or false negative and then another function could be passed that takes a confusion matrix and calculates a score based on that. I don't just want to optimize F1 scores, I also want to be able to optimize for F-beta scores, MCC, precision, recall and other metrics calculatable based on confusion matricies.

Another way would be to input a `sklearn.metrics` function and a function for normalising values into `True` and `False`. That would also very helpful.

**Comment by okhat** [comment]
https://dspy-docs.vercel.app/docs/building-blocks/metrics

**Comment by umarbutler** [comment]
@okhat This does not answer my question. Please see:
> The documentation states that '[f]or simple tasks, [a metric] could be just "accuracy" or "exact match" or "F1 score". This may be the case for simple classification or short-form QA tasks', yet it does not clarify how F1 scores can be used to optimize programs. Is that possible? And if so, how?
> The metric function that is passed to teleprompter.compile seems to take gold and pred as inputs, which are single classifications, so I am unable to see how you could calculate an F1 score based off of that. It would be really helpful to have that ability since not all tasks should be optimised based on accuracy.

**Comment by KevinGregory** [comment]
https://discord.com/channels/1161519468141355160/1161519642985111593/1234543339404267620

**Comment by chrico-bu-uab** [comment]
@umarbutler I too had a need for access to the full confusion matrix as well as scoring by a better performance metric (semantic f1 is effectively just confidence-weighted accuracy), so I made these two classes in the code below.

https://github.com/chrico-bu-uab/dspy_confusion/blob/main/dspy/evaluate/confusion.py

**Comment by fyears** [comment]
Hi I am also wondering how to compute the f1 score and optimize the prompt based on the f1 score, since I am doing a mult-label classification task. 

Is there any "standard" full example ? All the tutorials in the internet are using simple accuracy metric but it's essentially useful to use f1 in classifications.

I am guessing we have to use the param "trace" in validation function but no example is showing what it is and how to use it to get the summary metric of all sample. I am limited to the "one-by-one metric" aka accuracy.

Any help would be appreciated. Thanks everyone!
